compiler/src/iree/compiler/GlobalOptimization/DecomposeLowParallelismAttentionPass.cpp

```diff
@@ -0,0 +1,118 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/Dialect/LinalgExt/Utils/IndexingUtils.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+#define GEN_PASS_DEF_DECOMPOSELOWPARALLELISMATTENTIONPASS
+#include "iree/compiler/GlobalOptimization/Passes.h.inc"
+
+using namespace IREE::LinalgExt;
+
+namespace {
+
+const int64_t kDynamicDimMaxValue = 65536; // 2^16
+/// TODO: The M heuristic is a conservative choice I(Groverkss) made.
+/// We ideally need more data to determine when it is good to
+/// decompose.
+const int64_t kMParallelismThreshold = 4;
+/// The K1 heuristic is simply based on the fact that if we have a tile size
+/// bigger than 256, we should probably be tiling that dimension. Non decomposed
+/// attention is not tilable along K1 diomension.
+const int64_t kK1ParallelismThreshold = 256;
+
+static int64_t calculateSize(ArrayRef<int64_t> bounds, ArrayRef<int64_t> dims) {
+  int64_t size = 1;
+  for (int64_t dim : dims) {
+    if (ShapedType::isDynamic(bounds[dim])) {
+      // TODO: We should query range information to find the maximum value
+      // possible instead.
+      size *= kDynamicDimMaxValue;
+    } else {
+      size *= bounds[dim];
+    }
+  }
+  return size;
+}
+
+static bool hasLowParallelism(AttentionOp attnOp) {
+  FailureOr<AttentionOpDetail> opInfo =
+      AttentionOpDetail::get(attnOp.getIndexingMapsArray());
+
+  FailureOr<SmallVector<int64_t>> maybeBounds = attnOp.getStaticLoopRanges();
+  if (failed(maybeBounds)) {
+    return false;
+  }
+
+  SmallVector<int64_t> bounds = maybeBounds.value();
+
+  int64_t mSize = calculateSize(bounds, opInfo->getMDims());
+  int64_t k1Size = calculateSize(bounds, opInfo->getK1Dims());
+
+  // Fused Attention (aka Flash Attention) is parallelisable primarily over
+  // B, M dimensions. It is possible to parallelise over N dimension, but it
+  // leads to duplication of computation of the QK matmul and the softmax.
+  //
+  // If we do not have enough parallelism over M dimension, it is generally
+  // not useful to do fused attention, instead decomposing is a better option.
+  //
+  // Also, one of the main reasons for using fused attention is that decomposing
+  // leads to a tensor of size O(M x K2) as output, which generally grows
+  // as O(seq_len^2) and consumes a lot of memory. However, if M is small, we
+  // don't get much benefit on memory from doing fused attention.
+  //
+  // TODO: We should probably take batch size into account. Sometimes doing
+  // split-k gives you good parallelism, and we can be fine.
+  if (mSize <= kMParallelismThreshold) {
+    return true;
+  }
+
+  // The main problem of Fused Attention is that it reduces parallelism across
+  // K1, we cannot tile K1 in Fused Attention. This is generally okay, because
```

Comment:
What does "reduces parallelism" means here ? 
If we ignore, reduction-tree like or subgroup matmul k-reduction parallelism, practically there is no parallelism in K1. 

However, that does not imply we cannot tile K1 ?
We could actually tile K1 (the same way K2 is tiled) for larger K1 cases and do a iterative tiled accumulation. 

I understand that is not how its implemented. Correct me if I am wrong.
Maybe, then the docs should read as how it is implemented ?

Reply:
I think technically, we can have a nested scf.for to represent the tiled K1, but I think we may need to a little more work for that? But I guess the main Q here is, to decide if there an advantage of doing FA when K1 is really large or would it be more beneficial to have more "freedom" of tile sizes/schedule by having the 1st matmul in it's own dispatch

Reply:
In my experience, the decision of doing the FA or not is usually orthogonal on K1 -- in the sense if you do first matmul standalone K1 would still be a reduction dimension. 

FA trumps where M x K2 is larger and that bandwidth is what we generallly save.